#!/usr/bin/env python3

import os
import sys
import subprocess
import shutil
import re
import argparse

###########################
## USER ARGUMENT PARSING ##
###########################
parser = argparse.ArgumentParser(description='Read arguments')
parser.add_argument('-a', '--annotation', type=str, help='Annotation file for features of interest (gff format)')
parser.add_argument('-m', '--methylation', type=str, help='Methylation levels file (as generated by BSparse)')
parser.add_argument('-r', '--RNAseq', type=str, help='RNAseq alignment file (bam format)')
parser.add_argument('-p', '--piRNA', type=str, help='piRNA alignment file (bam format)')
parser.add_argument('-s', '--siRNA', type=str, help='siRNA alignment file (bam format)')
args = parser.parse_args()
# annotation file parsing
annotation_file = args.annotation
if annotation_file is not None:
	print("Annotation file is " + annotation_file)
else:
	print("ERROR: no annotation file (-a) specified")
	sys.exit(0)
# methylation file parsing
methylation_file = args.methylation
if methylation_file is not None:
	print("Methylation file is " + methylation_file)
else:
	print("ERROR: no methylation file (-m) specified")
	sys.exit(0)
# RNAseq file parsing
RNAseq_file = args.RNAseq
if RNAseq_file is not None:
	print("RNAseq file is " + RNAseq_file)
else:
	print("ERROR: no RNAseq file (-r) specified")
	sys.exit(0)
# piRNA file parsing
piRNA_file = args.piRNA
if piRNA_file is not None:
	print("piRNA file is " + piRNA_file)
else:
	print("ERROR: no piRNA file (-p) specified")
	sys.exit(0)
# siRNA file parsing
siRNA_file = args.siRNA
if siRNA_file is not None:
	print("siRNA file is " + siRNA_file)
else:
	print("ERROR: no siRNA file (-s) specified")
	sys.exit(0)

#################
## ACTUAL CODE ##
#################

